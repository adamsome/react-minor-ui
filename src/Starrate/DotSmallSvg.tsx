import React from 'react'

const DotSmallSvg = () => {
  return (
    <svg viewBox="0 0 512 512">
      <circle
        cx="256"
        cy="256"
        r="40"
        stroke-width="0"
        fill="currentColor"
        transform="inherit"
        transform-origin="center"
      />
    </svg>
  )
}

export default DotSmallSvg
