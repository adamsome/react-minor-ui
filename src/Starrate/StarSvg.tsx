import React from 'react'

const StarSvg = () => {
  return (
    <svg viewBox="0 0 51 48">
      <path
        fill="currentColor"
        stroke="currentColor"
        d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
      />
    </svg>
  )
}

export default StarSvg
