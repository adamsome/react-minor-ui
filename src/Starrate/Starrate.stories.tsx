import { action } from '@storybook/addon-actions'
import { boolean, number, withKnobs } from '@storybook/addon-knobs'
import React from 'react'
import { SIZES } from '../size'
import Starrate from './Starrate'

export default {
  component: Starrate,
  title: 'Starrate',
  excludeStories: ['actions'],
  decorators: [withKnobs],
}

export const actions = {
  onRatingChange: action('onRatingChange'),
}

const common = {
  ...actions,
  style: { marginBottom: '1rem' },
}

export const unrated = () => (
  <>
    {[...SIZES].reverse().map((size, i) => (
      <Starrate key={i} size={size} {...common} />
    ))}
  </>
)

export const rated = () => (
  <>
    {[...SIZES].reverse().map((size, i) => (
      <Starrate key={i} rating={number('rating', 3)} size={size} {...common} />
    ))}
  </>
)

export const maxRating = () => (
  <Starrate
    maxRating={number('maxRating', 10)}
    rating={number('rating', 3)}
    {...actions}
  />
)

export const disabled = () => (
  <Starrate
    disabled={boolean('disabled', true)}
    rating={number('rating', 3)}
    {...actions}
  />
)

export const colored = () => (
  <div style={{ color: '#0971f5' }}>
    <Starrate
      rating={number('rating', 3)}
      disabled={boolean('disabled', false)}
      {...actions}
    />
  </div>
)
