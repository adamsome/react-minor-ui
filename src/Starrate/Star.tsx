import React from 'react'
import { DEFAULT_SIZE_PROP, getIconWidth } from '../size'
import DotSmallSvg from './DotSmallSvg'
import StarSvg from './StarSvg'

export type StarProps = typeof defaultProps & {
  onMouseEnter?: () => void
  onMouseLeave?: () => void
  onClick?: () => void
}

export type StarState = 'off' | 'dim' | 'lit'

const defaultProps = {
  ...DEFAULT_SIZE_PROP,
  state: 'off' as StarState,
  dimOpacity: 0.25,
  disabled: false,
}

const Star = (props: StarProps) => {
  const { size, state, dimOpacity, disabled, ...mouseActions } = props

  const dotStyle: React.CSSProperties = {
    opacity: getDotOpacity(props),
    transition: 'opacity 100ms ease-in-out',
  }
  const starStyle: React.CSSProperties = {
    opacity: getStarOpacity(props),
    transform: `scale(${state === 'off' ? 0 : 1})`,
    transformOrigin: 'center',
    transition: 'opacity 200ms ease-in-out, transform 100ms ease-in-out',
  }

  const w = getIconWidth(size)
  const cursor = disabled ? 'auto' : 'pointer'

  return (
    <svg width={w} height={w} style={{ cursor }} {...mouseActions}>
      <g style={dotStyle} x={-0.2} y={0.9}>
        <DotSmallSvg />
      </g>
      <g style={starStyle}>
        <StarSvg />
      </g>
    </svg>
  )
}

const getDotOpacity = (props: StarProps): number => {
  switch (props.state) {
    case 'off':
      return props.dimOpacity
    case 'dim':
    case 'lit':
      return 0
  }
}

const getStarOpacity = (props: StarProps): number => {
  switch (props.state) {
    case 'off':
      return 0
    case 'dim':
      return props.dimOpacity
    case 'lit':
      return 1
  }
}

Star.defaultProps = defaultProps

export default Star
