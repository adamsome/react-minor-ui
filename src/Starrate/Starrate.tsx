import React, { useState } from 'react'
import { DEFAULT_SIZE_PROP } from '../size'
import Star, { StarState } from './Star'

type StarrateProps = typeof defaultProps & {
  rating?: number
  style?: React.CSSProperties
  onRatingChange?: (rating: number) => void
}

const defaultProps = {
  ...DEFAULT_SIZE_PROP,
  disabled: false,
  maxRating: 5,
}

const Starrate = (props: StarrateProps) => {
  const { rating, maxRating, size, disabled, style, onRatingChange } = props

  const [hoverIndex, setHoverIndex] = useState<number>(-1)

  const resetHover = () => setHoverIndex(-1)
  const setHover = (i: number) => () => setHoverIndex(i)

  const handleClick = (i: number) => () =>
    onRatingChange && onRatingChange(i + 1)

  const getStarState = (i: number): StarState => {
    const unlit = () => (rating != null ? 'dim' : 'off')

    if (!disabled && hoverIndex >= 0) {
      if (i <= hoverIndex) {
        return 'lit'
      }
      return unlit()
    }

    if (rating != null && i < rating) {
      return 'lit'
    }
    return unlit()
  }

  const createStar = (state: StarState, i: number): JSX.Element => (
    <Star
      key={i}
      state={state}
      disabled={disabled}
      size={size}
      onMouseEnter={setHover(i)}
      onMouseLeave={resetHover}
      onClick={handleClick(i)}
    />
  )

  const stars = range(0, maxRating)
    .map(getStarState)
    .map(createStar)

  return <div style={style}>{stars}</div>
}

const range = (from: number, to: number) => {
  const result = []
  let n = from
  while (n < to) {
    result.push(n)
    n += 1
  }
  return result
}

Starrate.defaultProps = defaultProps

export default Starrate
