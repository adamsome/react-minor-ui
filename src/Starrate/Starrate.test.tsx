import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Starrate from './Starrate'

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Starrate />, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders correctly', () => {
  const tree = renderer.create(<Starrate rating={3} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders unrated correctly', () => {
  const tree = renderer.create(<Starrate maxRating={2} />).toJSON()
  expect(tree).toMatchSnapshot()
})
