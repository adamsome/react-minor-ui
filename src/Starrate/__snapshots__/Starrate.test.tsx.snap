// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <svg
    height="1.5rem"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="1.5rem"
  >
    <g
      style={
        Object {
          "opacity": 0,
          "transition": "opacity 100ms ease-in-out",
        }
      }
      x={-0.2}
      y={0.9}
    >
      <svg
        viewBox="0 0 512 512"
      >
        <circle
          cx="256"
          cy="256"
          fill="currentColor"
          r="40"
          strokeWidth="0"
          transform="inherit"
          transform-origin="center"
        />
      </svg>
    </g>
    <g
      style={
        Object {
          "opacity": 1,
          "transform": "scale(1)",
          "transformOrigin": "center",
          "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
        }
      }
    >
      <svg
        viewBox="0 0 51 48"
      >
        <path
          d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          fill="currentColor"
          stroke="currentColor"
        />
      </svg>
    </g>
  </svg>
  <svg
    height="1.5rem"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="1.5rem"
  >
    <g
      style={
        Object {
          "opacity": 0,
          "transition": "opacity 100ms ease-in-out",
        }
      }
      x={-0.2}
      y={0.9}
    >
      <svg
        viewBox="0 0 512 512"
      >
        <circle
          cx="256"
          cy="256"
          fill="currentColor"
          r="40"
          strokeWidth="0"
          transform="inherit"
          transform-origin="center"
        />
      </svg>
    </g>
    <g
      style={
        Object {
          "opacity": 1,
          "transform": "scale(1)",
          "transformOrigin": "center",
          "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
        }
      }
    >
      <svg
        viewBox="0 0 51 48"
      >
        <path
          d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          fill="currentColor"
          stroke="currentColor"
        />
      </svg>
    </g>
  </svg>
  <svg
    height="1.5rem"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="1.5rem"
  >
    <g
      style={
        Object {
          "opacity": 0,
          "transition": "opacity 100ms ease-in-out",
        }
      }
      x={-0.2}
      y={0.9}
    >
      <svg
        viewBox="0 0 512 512"
      >
        <circle
          cx="256"
          cy="256"
          fill="currentColor"
          r="40"
          strokeWidth="0"
          transform="inherit"
          transform-origin="center"
        />
      </svg>
    </g>
    <g
      style={
        Object {
          "opacity": 1,
          "transform": "scale(1)",
          "transformOrigin": "center",
          "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
        }
      }
    >
      <svg
        viewBox="0 0 51 48"
      >
        <path
          d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          fill="currentColor"
          stroke="currentColor"
        />
      </svg>
    </g>
  </svg>
  <svg
    height="1.5rem"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="1.5rem"
  >
    <g
      style={
        Object {
          "opacity": 0,
          "transition": "opacity 100ms ease-in-out",
        }
      }
      x={-0.2}
      y={0.9}
    >
      <svg
        viewBox="0 0 512 512"
      >
        <circle
          cx="256"
          cy="256"
          fill="currentColor"
          r="40"
          strokeWidth="0"
          transform="inherit"
          transform-origin="center"
        />
      </svg>
    </g>
    <g
      style={
        Object {
          "opacity": 0.25,
          "transform": "scale(1)",
          "transformOrigin": "center",
          "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
        }
      }
    >
      <svg
        viewBox="0 0 51 48"
      >
        <path
          d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          fill="currentColor"
          stroke="currentColor"
        />
      </svg>
    </g>
  </svg>
  <svg
    height="1.5rem"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="1.5rem"
  >
    <g
      style={
        Object {
          "opacity": 0,
          "transition": "opacity 100ms ease-in-out",
        }
      }
      x={-0.2}
      y={0.9}
    >
      <svg
        viewBox="0 0 512 512"
      >
        <circle
          cx="256"
          cy="256"
          fill="currentColor"
          r="40"
          strokeWidth="0"
          transform="inherit"
          transform-origin="center"
        />
      </svg>
    </g>
    <g
      style={
        Object {
          "opacity": 0.25,
          "transform": "scale(1)",
          "transformOrigin": "center",
          "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
        }
      }
    >
      <svg
        viewBox="0 0 51 48"
      >
        <path
          d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          fill="currentColor"
          stroke="currentColor"
        />
      </svg>
    </g>
  </svg>
</div>
`;

exports[`renders unrated correctly 1`] = `
<div>
  <svg
    height="1.5rem"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="1.5rem"
  >
    <g
      style={
        Object {
          "opacity": 0.25,
          "transition": "opacity 100ms ease-in-out",
        }
      }
      x={-0.2}
      y={0.9}
    >
      <svg
        viewBox="0 0 512 512"
      >
        <circle
          cx="256"
          cy="256"
          fill="currentColor"
          r="40"
          strokeWidth="0"
          transform="inherit"
          transform-origin="center"
        />
      </svg>
    </g>
    <g
      style={
        Object {
          "opacity": 0,
          "transform": "scale(0)",
          "transformOrigin": "center",
          "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
        }
      }
    >
      <svg
        viewBox="0 0 51 48"
      >
        <path
          d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          fill="currentColor"
          stroke="currentColor"
        />
      </svg>
    </g>
  </svg>
  <svg
    height="1.5rem"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="1.5rem"
  >
    <g
      style={
        Object {
          "opacity": 0.25,
          "transition": "opacity 100ms ease-in-out",
        }
      }
      x={-0.2}
      y={0.9}
    >
      <svg
        viewBox="0 0 512 512"
      >
        <circle
          cx="256"
          cy="256"
          fill="currentColor"
          r="40"
          strokeWidth="0"
          transform="inherit"
          transform-origin="center"
        />
      </svg>
    </g>
    <g
      style={
        Object {
          "opacity": 0,
          "transform": "scale(0)",
          "transformOrigin": "center",
          "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
        }
      }
    >
      <svg
        viewBox="0 0 51 48"
      >
        <path
          d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          fill="currentColor"
          stroke="currentColor"
        />
      </svg>
    </g>
  </svg>
</div>
`;
